rust/src/ffi/exported.rs

```diff
@@ -0,0 +1,244 @@
+    state: *mut c_void,
+    out: *mut *mut c_char,
+    out_length: *mut u64,
+) {
+    unimplemented!()
+}
+
+/// Releases the buffer returned by GetMemoryFootprint.
+#[unsafe(no_mangle)]
+unsafe extern "C" fn Carmen_Rust_ReleaseMemoryFootprintBuffer(buf: *mut c_char, buf_length: u64) {
+    unimplemented!()
+}
+
+/// Compile-time check that the signatures of the exported functions match the ones generated from
+/// the C header file.
+#[allow(unused)]
+const COMPILE_TIME_CHECK_THAT_SIGNATURES_MATCH_SIGNATURES_GENERATED_FROM_C_HEADER: () = {
+    assert_same_signature(
+        Carmen_Rust_OpenState as unsafe extern "C" fn(_, _, _, _, _) -> _,
+        bindings::Carmen_Rust_OpenState,
+    );
+    assert_same_signature(
+        Carmen_Rust_GetArchiveState as unsafe extern "C" fn(_, _) -> _,
+        bindings::Carmen_Rust_GetArchiveState,
+    );
+    assert_same_signature(
+        Carmen_Rust_Flush as unsafe extern "C" fn(_),
+        bindings::Carmen_Rust_Flush,
+    );
+    assert_same_signature(
+        Carmen_Rust_Close as unsafe extern "C" fn(_),
+        bindings::Carmen_Rust_Close,
+    );
+    assert_same_signature(
+        Carmen_Rust_ReleaseState as unsafe extern "C" fn(_),
+        bindings::Carmen_Rust_ReleaseState,
+    );
+    assert_same_signature(
+        Carmen_Rust_GetAccountState as unsafe extern "C" fn(_, _, _),
+        bindings::Carmen_Rust_GetAccountState,
+    );
+    assert_same_signature(
+        Carmen_Rust_GetBalance as unsafe extern "C" fn(_, _, _),
+        bindings::Carmen_Rust_GetBalance,
+    );
+    assert_same_signature(
+        Carmen_Rust_GetNonce as unsafe extern "C" fn(_, _, _),
+        bindings::Carmen_Rust_GetNonce,
+    );
+    assert_same_signature(
+        Carmen_Rust_GetStorageValue as unsafe extern "C" fn(_, _, _, _),
+        bindings::Carmen_Rust_GetStorageValue,
+    );
+    assert_same_signature(
+        Carmen_Rust_GetCode as unsafe extern "C" fn(_, _, _, _),
+        bindings::Carmen_Rust_GetCode,
+    );
+    assert_same_signature(
+        Carmen_Rust_GetCodeHash as unsafe extern "C" fn(_, _, _),
+        bindings::Carmen_Rust_GetCodeHash,
+    );
+    assert_same_signature(
+        Carmen_Rust_GetCodeSize as unsafe extern "C" fn(_, _, _),
+        bindings::Carmen_Rust_GetCodeSize,
+    );
+    assert_same_signature(
+        Carmen_Rust_Apply as unsafe extern "C" fn(_, _, _, _),
+        bindings::Carmen_Rust_Apply,
+    );
+    assert_same_signature(
+        Carmen_Rust_GetHash as unsafe extern "C" fn(_, _),
+        bindings::Carmen_Rust_GetHash,
+    );
+    assert_same_signature(
+        Carmen_Rust_GetMemoryFootprint as unsafe extern "C" fn(_, _, _),
+        bindings::Carmen_Rust_GetMemoryFootprint,
+    );
+    assert_same_signature(
+        Carmen_Rust_ReleaseMemoryFootprintBuffer as unsafe extern "C" fn(_, _),
+        bindings::Carmen_Rust_ReleaseMemoryFootprintBuffer,
+    );
+};
+
+const fn assert_same_signature<T>(a: T, b: T) -> (T, T) {
```

Comment:
I was wondering whether the return type is even needed: It is, but I have no idea why :shrug:

Reply:
The problem is that we take T by value. T might implement `Drop`. But `Drop::drop` is not a `const fn`. Therefore, we can not drop `T` in const contexts. But returning it is fine because then it is again the callers' responsibility to drop T. At the call side we know the specific type of `T` (function pointers) and know that `T` does not implement `Drop`.
Another option would be to take `T` by reference, but this is uglier on the caller side.